### PR TITLE
[h2olog] use original headers instead of h2olog/quic.h

### LINKED ISSUE #2365: [h2olog] use original header files
At the moment, h2olog uses the following approach:
* define copies of certain structs (e.g., `quicly_conn_t`) in h2olog/quic.h
* dump all those struct members

As it is difficult and error-prone to have structs defined in two places, we need to change h2olog to use the header files that we use for building h2o, and in the source code of h2olog (or maybe in the doc-comment of *-probes.d), designate the fields that have to be logged.

### PR DESCRIPTION
**Pull request by gfx**
Resolve #2365

To give the BPF program with type information of h2o,  h2olog generates a part of the BPF program dynamically. It means `gen-quic-bpf.py` generates a code generator in C++.

It currently depends on the gcc / clang compiler extension: `sizeof`  

---

`gen_quic_bpf_header` generates something like this (the actual one is available with `-dd` option):

```c
#define sizeof_st_quicly_stream_t 128
/* st_quicly_stream_t__stream_id (struct st_quicly_stream_t#stream_id) */
#define offsetof_st_quicly_stream_t__stream_id 8
#define typeof_st_quicly_stream_t__stream_id int64_t
#define get_st_quicly_stream_t__stream_id(st) *((const int64_t *) ((const char*)st + offsetof_st_quicly_stream_t__stream_id))
```

---

At first, I tried `pahole` but it was difficult because `pahole`'s features are limited and I have to add a lot of heuristic logics to extract information from `pahole`s output. I believe its maintenance is much harder than it seems to be.

### CODE REVIEW

File: h2olog/misc/gen-quic-bpf.py [hunk 1]
```diff
@@ -257,10 +270,75 @@ def prepare_context(h2o_dir):
...
+
+def build_typedef_for_cplusplus():
+  typedef = st_quicly_conn_t_def
+
+  # CAUTION: it depends on a GCC/Clang compiler extension: typeof()
+  i = 0
+  for st_name, st_fields in struct_map.items():
+    for st_field_access, st_field_name_alias in st_fields:
+      i += 1
+      typedef += """static struct %s s%d; // placeholder for typeof()\n""" % (st_name, i)
```
Comment (kazuho) [inline h2olog/misc/gen-quic-bpf.py:336]: I think we can get rid of this line. The type can be obtained as `typeof(((type *)NULL)->field)`.

Comment (gfx) [reply h2olog/misc/gen-quic-bpf.py]: Actually! The same technique is used to calculate offsets.

Fixed 9a36c0c398c0e4459e7f710479ebac2f9e21fd5b

File: h2olog/misc/gen-quic-bpf.py [hunk 2]
```diff
@@ -30,7 +30,26 @@
 
 quicly_probes_d = "deps/quicly/quicly-probes.d"
 h2o_probes_d = "h2o-probes.d"
-data_types_h = "h2olog/quic.h"
+
+struct_map = {
+    # deps/quicly/include/quicly.h
+    "st_quicly_stream_t": [
+        ("stream_id", None),
```
[the remark this answers is not in the document]
Comment (gfx) [reply h2olog/misc/gen-quic-bpf.py:41]: Added 469f0ea343eb8b294edd453ef8dd596bc1749429

**Comment by gfx** [comment]
FYI https://github.com/h2o/h2o/pull/2367 is required to build this branch with old `cmake`.

**Pull request merged**